pkg/controller/syncmachineset/syncmachineset_controller.go

```diff
@@ -0,0 +1,548 @@
+		msLog.Debugf("error looking up machineset's cluster: %v", err)
+		return
+	}
+
+	machineSets, err := controller.MachineSetsForCluster(coCluster, c.machineSetsLister)
+	if err != nil {
+		logging.WithCluster(c.logger, coCluster).Debugf("error looking up cluster's machinesets: %v", err)
+		return
+	}
+	for _, ms := range machineSets {
+		logging.WithMachineSet(c.logger, ms).Debugf("Updating machine set")
+		c.enqueueMachineSet(ms)
+	}
+}
+
+// Run runs c; will not return until stopCh is closed. workers determines how
+// many machine sets will be handled in parallel.
+func (c *Controller) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer c.queue.ShutDown()
+
+	c.logger.Infof("Starting syncmachineset controller")
+	defer c.logger.Infof("Shutting down syncmachineset controller")
+
+	if !controller.WaitForCacheSync("syncmachineset", stopCh, c.machineSetsSynced) {
+		c.logger.Errorf("could not sync caches for syncmachineset controller")
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(c.worker, time.Second, stopCh)
+	}
+
+	<-stopCh
+}
+
+func (c *Controller) enqueue(machineSet *cov1.MachineSet) {
+	key, err := controller.KeyFunc(machineSet)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", machineSet, err))
+		return
+	}
+
+	c.queue.Add(queuedMachineSet{
+		key: key,
+	})
+}
+
+// enqueueAfter will enqueue a machine set after the provided amount of time.
+func (c *Controller) enqueueAfter(ms *cov1.MachineSet, after time.Duration) {
+	key, err := controller.KeyFunc(ms)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("couldn't get key for object %#v: %v", ms, err))
+		return
+	}
+
+	c.queue.AddAfter(queuedMachineSet{
+		key: key,
+	}, after)
+}
+
+// worker runs a worker thread that just dequeues items, processes them, and marks them done.
+// It enforces that the syncHandler is never invoked concurrently with the same key.
+func (c *Controller) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *Controller) processNextWorkItem() bool {
+	queuedMS, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(queuedMS)
+
+	err := c.syncHandler(queuedMS.(queuedMachineSet))
+	c.handleErr(err, queuedMS)
+
+	return true
+}
+
+func (c *Controller) handleErr(err error, queuedMS interface{}) {
+	if err == nil {
+		c.queue.Forget(queuedMS)
+		return
+	}
+
+	logger := c.logger.WithField("machineset", queuedMS)
+
+	if c.queue.NumRequeues(queuedMS) < maxRetries {
+		logger.Infof("error syncing machine set: %v", err)
+		c.queue.AddRateLimited(queuedMS)
+		return
+	}
+
+	utilruntime.HandleError(err)
+	logger.Infof("dropping machine set out of the queue: %v", err)
+	c.queue.Forget(queuedMS)
+}
+
+func isMasterMachineSet(machineSet *cov1.MachineSet) bool {
+	return machineSet.Spec.NodeType == cov1.NodeTypeMaster
+}
+
+func (c *Controller) buildRemoteClusterClients(machineSet *cov1.MachineSet) (clusterapiclient.Interface, error) {
```

Comment:
We should extract this function since it is doing the same thing as this function. We can do that as follow-on work, though.

https://github.com/openshift/cluster-operator/blob/1d8612326880b51296b3fee44537c3d78abde5e6/pkg/controller/cluster/cluster_controller.go#L367